v2/report_publisher.go

```diff
@@ -1,27 +1,106 @@
 package bugsnag
 
-import "fmt"
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+)
 
 type reportPublisher interface {
 	publishReport(*payload) error
+	setMainProgramContext(context.Context)
 }
 
-type defaultReportPublisher struct{}
+var mainProgramDone = false
 
-func (*defaultReportPublisher) publishReport(p *payload) error {
+func (defPub *defaultReportPublisher) delivery() {
+	signalsCh := make(chan os.Signal, 1)
+	signal.Notify(signalsCh, syscall.SIGINT, syscall.SIGTERM)
+	closeDelivery := false
+
+waitForEnd:
+	for {
+		select {
+		case <-signalsCh:
+			mainProgramDone = true
+			break waitForEnd
+		case <-defPub.mainProgramCtx.Done():
+			mainProgramDone = true
+			break waitForEnd
+		case p, ok := <-defPub.eventsChan:
+			if ok {
+				if err := p.deliver(); err != nil {
+					// Ensure that any errors are logged if they occur in a goroutine.
+					p.logf("bugsnag/defaultReportPublisher.publishReport: %v", err)
+				}
+			} else {
+				p.logf("Event channel closed")
+				return
+			}
+		}
+	}
+
+	// Send remaining elements from the queue
+	for !closeDelivery {
+		select {
+		case p, ok := <-defPub.eventsChan:
+			if ok {
+				if err := p.deliver(); err != nil {
+					// Ensure that any errors are logged if they occur in a goroutine.
+					p.logf("bugsnag/defaultReportPublisher.publishReport: %v", err)
+				}
+			} else {
+				p.logf("Event channel closed")
+				return
+			}
+		default:
+			if mainProgramDone && len(defPub.eventsChan) == 0 {
+				fmt.Println("Main program done and event channel empty")
+				closeDelivery = true
+			}
+		}
+	}
+}
+
+type defaultReportPublisher struct{
+	eventsChan chan *payload
+	mainProgramCtx context.Context
+}
+
+func NewPublisher() *defaultReportPublisher {
+	defPub := defaultReportPublisher{}
+	defPub.eventsChan = make(chan *payload, 100)
+	defPub.mainProgramCtx = context.TODO()
+
+	go defPub.delivery()
+
+	return &defPub
+}
+
+func (defPub *defaultReportPublisher) setMainProgramContext(ctx context.Context) {
```

Comment:
Does this need to be done as a separate step? Can it be part of the constructor?

Reply:
as discussed - this should be a separate step because of the order of creation, publisher is created as soon as module loads but user context could only be provided later, at Configure stage